File: python/copilot/_mode.py
"""
Mode = "empty" support: ToolSet builder, BUILTIN_TOOLS_ISOLATED, and helpers
that translate Mode = "empty" into runtime-level session options.

The runtime is mode-agnostic; the SDK is what turns ``mode="empty"`` into the
right combination of options on the wire (no environment_context, telemetry
off, custom instructions off, etc.). Callers can opt back in field-by-field.
"""

from __future__ import annotations

import re
from collections.abc import Iterable
from typing import Any, Literal

CopilotClientMode = Literal["copilot-cli", "empty"]

_TOOL_NAME_REGEX = re.compile(r"^[a-zA-Z0-9_-]+$")


def _validate_tool_name(kind: str, name: str) -> None:
    if not name:
        raise ValueError(f"invalid {kind} tool name: must not be empty")
    if name == "*":
        return
    if not _TOOL_NAME_REGEX.match(name):
        raise ValueError(
            f"invalid {kind} tool name {name!r}: tool names must match "
            r"/^[a-zA-Z0-9_-]+$/ or be the wildcard '*'"
        )


class ToolSet:
    """Builder for source-qualified tool filter patterns.

    ``ToolSet`` accumulates entries like ``builtin:bash``, ``mcp:*``, or
    ``custom:my_tool`` for use in
    :class:`CopilotClient.create_session`'s ``available_tools`` /
    ``excluded_tools`` parameters.

    Tool classification (``builtin``/``mcp``/``custom``) is determined by the
    runtime at registration time — not by name parsing — so
    ``add_builtin("foo")`` only matches tools the runtime registered as
    built-in.
    """

    def __init__(self) -> None:
        self._items: list[str] = []

    def add_builtin(self, name: str | Iterable[str]) -> ToolSet:
        """Add a built-in tool pattern (``"bash"``/``"*"``/an iterable of names)."""
        if isinstance(name, str):
            _validate_tool_name("builtin", name)
            self._items.append(f"builtin:{name}")
        else:
            for n in name:
                _validate_tool_name("builtin", n)
                self._items.append(f"builtin:{n}")
        return self

    def add_custom(self, name: str) -> ToolSet:
        """Add a custom-tool pattern (e.g. ``"my_tool"`` or ``"*"``)."""
        _validate_tool_name("custom", name)
        self._items.append(f"custom:{name}")
        return self

    def add_mcp(self, tool_name: str) -> ToolSet:
        """Add an MCP tool pattern (e.g. ``"github-list_issues"`` or ``"*"``)."""
        _validate_tool_name("mcp", tool_name)
        self._items.append(f"mcp:{tool_name}")
        return self

    def to_list(self) -> list[str]:
        """Return a defensive copy of the accumulated filter strings."""
        return list(self._items)

    def __iter__(self):
        return iter(self.to_list())

    def __len__(self) -> int:
        return len(self._items)


#: Built-in tools that operate only within a single session — no host FS
#: access outside the session, no cross-session state, no host environment
#: access, no network. Safe to enable in ``mode="empty"`` scenarios without
#: leaking host capabilities.
#:
#: Contract: tools in this set MUST NOT be extended (even behind options or
#: args) to read or write state outside the session boundary. Adding
#: cross-session or host-state behavior to one of these tools is a breaking
#: change that requires removing it from this set.
BUILTIN_TOOLS_ISOLATED: list[str] = [
    "ask_user",
    "task_complete",
    "exit_plan_mode",
    "task",
    "read_agent",
    "write_agent",
    "list_agents",
    "send_inbox",
    "context_board",
    "skill",
]


def _normalize_tool_filter(value: Any) -> list[str] | None:
    """Accept ``ToolSet``, ``list[str]``, or ``None``; return a list or ``None``.

    Reject plain ``str`` explicitly — ``list("foo")`` would silently shred it
    into characters, sending an invalid tool filter list on the wire.
    """
    if value is None:
        return None
    if isinstance(value, ToolSet):
        return value.to_list()
    if isinstance(value, str):
        raise TypeError(
            "tool filter must be a ToolSet or list[str], not str. "
            'Pass a single-element list (e.g. ["builtin:bash"]) or a '
            "ToolSet (e.g. ToolSet().add_builtin('bash'))."
        )
    return list(value)


def _validate_tool_filter_list(field: str, items: list[str] | None) -> None:
    """Reject bare ``"*"`` entries (must use ``builtin:*``/``mcp:*``/``custom:*``)."""
    if items is None:
        return
    for entry in items:
        if entry == "*":
            raise ValueError(
                f"invalid {field} entry '*': there is no bare wildcard. "
                "Use ToolSet().add_builtin('*'), .add_mcp('*'), or "
                ".add_custom('*') to target a specific source."
            )


def _system_message_for_mode(
    mode: CopilotClientMode | None,
    supplied: Any,
) -> Any:
    """Apply empty-mode environment_context stripping to a system message dict.

    The caller passes the already-normalized wire payload (a ``dict`` with
    ``mode`` / ``content`` / ``sections``) or ``None``. The caller's value
    wins if it already specifies an ``environment_context`` override.
    """
    if mode != "empty":
        return supplied
    remove_action = {"action": "remove"}
    if supplied is None:
        return {"mode": "customize", "sections": {"environment_context": remove_action}}
    supplied_mode = supplied.get("mode", "")
    if supplied_mode == "replace":
        return supplied
    if supplied_mode == "customize":
        sections = supplied.get("sections") or {}
        if "environment_context" in sections:
            return supplied
        merged = {**supplied, "sections": {**sections, "environment_context": remove_action}}
        return merged
    # append (or unspecified): promote to customize so we can also strip
    # environment_context. The runtime appends ``content`` in both modes, so
    # the caller's text is preserved verbatim.
    out: dict[str, Any] = {
        "mode": "customize",
        "sections": {"environment_context": remove_action},
    }
    if "content" in supplied and supplied["content"] is not None:
        out["content"] = supplied["content"]
    return out


def _empty_mode_bool_default(
    mode: CopilotClientMode | None,
    supplied: bool | None,
    empty_default: bool,
) -> bool | None:
    if mode == "empty" and supplied is None:
        return empty_default
    return supplied


def _enable_session_telemetry_default(
    mode: CopilotClientMode | None,
    supplied: bool | None,
) -> bool | None:
    """Empty mode defaults telemetry to False; caller value wins."""
    return _empty_mode_bool_default(mode, supplied, False)


def _skip_embedding_retrieval_default(
    mode: CopilotClientMode | None,
    supplied: bool | None,
) -> bool | None:
    """Empty mode defaults embedding retrieval to off; caller value wins."""
    return _empty_mode_bool_default(mode, supplied, True)


def _embedding_cache_storage_default(
    mode: CopilotClientMode | None,
    supplied: Literal["persistent", "in-memory"] | None,
) -> Literal["persistent", "in-memory"] | None:
    """Empty mode defaults embedding cache storage to in-memory; caller value wins."""
    if mode == "empty" and supplied is None:
        return "in-memory"
    return supplied


def _enable_on_demand_instruction_discovery_default(
    mode: CopilotClientMode | None,
    supplied: bool | None,
) -> bool | None:
    """Empty mode defaults on-demand instruction discovery to False."""
    return _empty_mode_bool_default(mode, supplied, False)


def _enable_file_hooks_default(
    mode: CopilotClientMode | None,
    supplied: bool | None,
) -> bool | None:
    """Empty mode defaults file hooks to False; caller value wins."""
    return _empty_mode_bool_default(mode, supplied, False)


def _enable_host_git_operations_default(
    mode: CopilotClientMode | None,
    supplied: bool | None,
) -> bool | None:
    """Empty mode defaults host git operations to False; caller value wins."""
    return _empty_mode_bool_default(mode, supplied, False)


def _enable_session_store_default(
    mode: CopilotClientMode | None,
    supplied: bool | None,
) -> bool | None:
    """Empty mode defaults the session store to False; caller value wins."""
    return _empty_mode_bool_default(mode, supplied, False)


def _enable_skills_default(
    mode: CopilotClientMode | None,
    supplied: bool | None,
) -> bool | None:
    """Empty mode defaults skills to False; caller value wins."""
    return _empty_mode_bool_default(mode, supplied, False)


def _mcp_oauth_token_storage_default(
    mode: CopilotClientMode | None,
    supplied: Literal["persistent", "in-memory"] | None,
) -> Literal["persistent", "in-memory"] | None:
    """Empty mode defaults MCP OAuth token storage to in-memory; caller value wins."""
    if mode == "empty" and supplied is None:
        return "in-memory"
    return supplied


def _post_create_options_patch(
    mode: CopilotClientMode | None,
    skip_custom_instructions: bool | None,
    custom_agents_local_only: bool | None,
    coauthor_enabled: bool | None,
    manage_schedule_enabled: bool | None,
) -> dict[str, Any] | None:
    """Build the patch sent via ``session.options.update`` after create/resume.

    In empty mode the four overridable flags default to safe values
    (caller-supplied values win); ``installedPlugins=[]`` is unconditional.
    Returns ``None`` if no patch should be sent.
    """
    if mode == "empty":
        patch: dict[str, Any] = {
            "skipCustomInstructions": (
                skip_custom_instructions if skip_custom_instructions is not None else True
            ),
            "customAgentsLocalOnly": (
                custom_agents_local_only if custom_agents_local_only is not None else True
            ),
            "coauthorEnabled": coauthor_enabled if coauthor_enabled is not None else False,
            "manageScheduleEnabled": (
                manage_schedule_enabled if manage_schedule_enabled is not None else False
            ),
            "installedPlugins": [],
        }
        return patch
    patch = {}
    if skip_custom_instructions is not None:
        patch["skipCustomInstructions"] = skip_custom_instructions
    if custom_agents_local_only is not None:
        patch["customAgentsLocalOnly"] = custom_agents_local_only
    if coauthor_enabled is not None:
        patch["coauthorEnabled"] = coauthor_enabled
    if manage_schedule_enabled is not None:
        patch["manageScheduleEnabled"] = manage_schedule_enabled
    return patch or None


def _require_storage_for_empty_mode(
    *,
    mode: CopilotClientMode | None,
    base_directory: str | None,
    session_fs_set: bool,
    is_uri_connection: bool,
) -> None:
    if mode != "empty":
        return
    if base_directory or session_fs_set or is_uri_connection:
        return
    raise ValueError(
        "CopilotClient(mode='empty') requires base_directory, session_fs, "
        "or a UriRuntimeConnection. Empty mode needs explicit per-tenant "
        "storage and won't fall back to ~/.copilot."
    )


def _require_available_tools_for_empty_mode(
    mode: CopilotClientMode | None,
    available_tools: list[str] | None,
) -> None:
    if mode == "empty" and available_tools is None:
        raise ValueError(
            "CopilotClient is in mode='empty' but create_session was called "
            "without available_tools. Empty mode requires every session to "
            "explicitly opt into the tools it wants — e.g. "
            "ToolSet().add_builtin(BUILTIN_TOOLS_ISOLATED)."
        )

File: rust/src/mode.rs
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

//! Client-level "empty" mode for minimal/safe defaults.
//!
//! See the plan in <https://github.com/github/copilot-agent-runtime/issues/7155>:
//! [`ClientMode::Empty`] disables ambient CLI-style behavior by default so an
//! app must explicitly opt back into features. This module exposes the public
//! enum, the [`ToolSet`] builder for source-qualified tool filter patterns,
//! and the [`BUILTIN_TOOLS_ISOLATED`] curated allowlist.

use std::collections::HashMap;

use crate::types::{MemoryConfiguration, SectionOverride, SystemMessageConfig};

/// Controls SDK defaults for ambient CLI-style behavior.
///
/// - [`ClientMode::CopilotCli`] (default): defaults equivalent to Copilot CLI.
///   Useful when building a coding agent that shares sessions with Copilot CLI.
///   **Do not use this mode for server-based multi-user applications** — the
///   default coding agent has tools and capabilities that operate across
///   sessions and can access the host OS environment.
/// - [`ClientMode::Empty`]: disables optional features by default. The app
///   must explicitly opt into anything it needs. Required for any scenario
///   where CLI-like ambient behavior is unsafe (e.g. multi-user servers).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum ClientMode {
    /// Defaults equivalent to Copilot CLI (the default).
    #[default]
    CopilotCli,
    /// Disables optional features by default; app must opt in explicitly.
    Empty,
}

/// Tool name character set enforced by the runtime at every registration
/// boundary. Mirrors the runtime's `VALID_TOOL_NAME_REGEX`.
fn is_valid_tool_name(name: &str) -> bool {
    !name.is_empty()
        && name
            .chars()
            .all(|c| c.is_ascii_alphanumeric() || c == '_' || c == '-')
}

fn validate_name(kind: &str, name: &str) -> Result<(), crate::Error> {
    if name == "*" {
        return Ok(());
    }
    if !is_valid_tool_name(name) {
        return Err(crate::Error::with_message(
            crate::ErrorKind::InvalidConfig,
            format!(
                "Invalid {kind} tool name '{name}': tool names must match \
             /^[a-zA-Z0-9_-]+$/ or be the wildcard '*'."
            ),
        ));
    }
    Ok(())
}

/// Builder that produces source-qualified tool filter strings (e.g.
/// `"builtin:bash"`, `"mcp:*"`, `"custom:foo"`) for the session's
/// `available_tools` list.
///
/// Tools are classified by the runtime at registration time, not from name
/// parsing — so `add_builtin("foo")` matches only tools registered as
/// built-in, even if an MCP server happens to register a tool with the same
/// wire name.
///
/// # Example
///
/// ```
/// # use github_copilot_sdk::mode::{ToolSet, BUILTIN_TOOLS_ISOLATED};
/// let tools = ToolSet::new()
///     .add_builtin_many(BUILTIN_TOOLS_ISOLATED)?
///     .add_mcp("*")?
///     .add_custom("*")?
///     .to_vec();
/// # Ok::<(), github_copilot_sdk::Error>(())
/// ```
#[derive(Debug, Clone, Default)]
pub struct ToolSet {
    items: Vec<String>,
}

impl ToolSet {
    /// Construct an empty tool set.
    pub fn new() -> Self {
        Self::default()
    }

    /// Add a single built-in tool pattern. Pass a specific name (e.g.
    /// `"bash"`) or `"*"` to match all built-in tools.
    pub fn add_builtin(mut self, name: &str) -> Result<Self, crate::Error> {
        validate_name("builtin", name)?;
        self.items.push(format!("builtin:{name}"));
        Ok(self)
    }

    /// Add a list of built-in tool patterns (e.g. [`BUILTIN_TOOLS_ISOLATED`]).
    pub fn add_builtin_many<I, S>(mut self, names: I) -> Result<Self, crate::Error>
    where
        I: IntoIterator<Item = S>,
        S: AsRef<str>,
    {
        for name in names {
            let name = name.as_ref();
            validate_name("builtin", name)?;
            self.items.push(format!("builtin:{name}"));
        }
        Ok(self)
    }

    /// Add a custom tool pattern. Matches tools registered via the SDK's
    /// `tools` option or via custom agents.
    pub fn add_custom(mut self, name: &str) -> Result<Self, crate::Error> {
        validate_name("custom", name)?;
        self.items.push(format!("custom:{name}"));
        Ok(self)
    }

    /// Add an MCP tool pattern. Pass the runtime's canonical wire name
    /// (e.g. `"github-list_issues"`) or `"*"` to match all MCP tools.
    pub fn add_mcp(mut self, tool_name: &str) -> Result<Self, crate::Error> {
        validate_name("mcp", tool_name)?;
        self.items.push(format!("mcp:{tool_name}"));
        Ok(self)
    }

    /// Returns a defensive copy of the accumulated filter strings.
    pub fn to_vec(&self) -> Vec<String> {
        self.items.clone()
    }

    /// Returns the accumulated filter strings, consuming the builder.
    pub fn into_vec(self) -> Vec<String> {
        self.items
    }

    /// Number of accumulated filter strings.
    pub fn len(&self) -> usize {
        self.items.len()
    }

    /// Returns `true` if no filter strings have been added.
    pub fn is_empty(&self) -> bool {
        self.items.is_empty()
    }
}

impl From<ToolSet> for Vec<String> {
    fn from(value: ToolSet) -> Self {
        value.into_vec()
    }
}

/// Built-in tools that operate only within the bounds of a single session —
/// no host filesystem access outside the session, no cross-session state,
/// no host environment access, no network.
///
/// Safe to enable in [`ClientMode::Empty`] scenarios (e.g. multi-tenant
/// servers) without leaking host capabilities.
///
/// **Contract:** tools in this set MUST NOT be extended (even behind options
/// or args) to read or write state outside the session boundary. Adding
/// cross-session or host-state behavior to one of these tools is a breaking
/// change that requires removing it from this set.
pub const BUILTIN_TOOLS_ISOLATED: &[&str] = &[
    "ask_user",
    "task_complete",
    "exit_plan_mode",
    "task",
    "read_agent",
    "write_agent",
    "list_agents",
    "send_inbox",
    "context_board",
    "skill",
];

/// Validate a tool filter list (`available_tools` or `excluded_tools`).
/// Rejects the bare `"*"` shorthand with a clear error pointing the developer
/// at the source-qualified forms.
pub(crate) fn validate_tool_filter_list(
    field: &str,
    list: Option<&[String]>,
) -> Result<(), crate::Error> {
    let Some(list) = list else { return Ok(()) };
    for item in list {
        if item == "*" {
            return Err(crate::Error::with_message(
                crate::ErrorKind::InvalidConfig,
                format!(
                    "{field} contains a bare '*' which matches no tool. Use \
                 source-qualified wildcards instead: \
                 ToolSet::new().add_builtin(\"*\").add_mcp(\"*\").add_custom(\"*\")."
                ),
            ));
        }
    }
    Ok(())
}

/// Returns the system message config to use, adjusted for the current mode.
/// In empty mode we ensure the `environment_context` section is removed
/// unless the app has already taken control of it.
pub(crate) fn system_message_for_mode(
    mode: ClientMode,
    supplied: Option<SystemMessageConfig>,
) -> Option<SystemMessageConfig> {
    if mode != ClientMode::Empty {
        return supplied;
    }
    let strip_env = || {
        let mut sections = HashMap::new();
        sections.insert(
            "environment_context".to_string(),
            SectionOverride {
                action: Some("remove".to_string()),
                content: None,
            },
        );
        sections
    };
    let Some(supplied) = supplied else {
        return Some(SystemMessageConfig {
            mode: Some("customize".to_string()),
            content: None,
            sections: Some(strip_env()),
        });
    };
    let mode_str = supplied.mode.as_deref().unwrap_or("append");
    match mode_str {
        "replace" => Some(supplied),
        "customize" => {
            if supplied
                .sections
                .as_ref()
                .is_some_and(|s| s.contains_key("environment_context"))
            {
                Some(supplied)
            } else {
                let mut sections = supplied.sections.unwrap_or_default();
                sections.insert(
                    "environment_context".to_string(),
                    SectionOverride {
                        action: Some("remove".to_string()),
                        content: None,
                    },
                );
                Some(SystemMessageConfig {
                    mode: Some("customize".to_string()),
                    content: supplied.content,
                    sections: Some(sections),
                })
            }
        }
        // "append" or any unrecognized value: promote to customize so we
        // can also strip environment_context; the runtime appends `content`
        // to additional instructions either way.
        _ => Some(SystemMessageConfig {
            mode: Some("customize".to_string()),
            content: supplied.content,
            sections: Some(strip_env()),
        }),
    }
}

/// Returns the memory configuration to use, adjusted for the current mode.
///
/// In [`ClientMode::Empty`] the memory feature defaults to disabled so an app
/// must opt in explicitly. In [`ClientMode::CopilotCli`] no SDK default is
/// applied: the configuration is left unset so the runtime applies its own
/// default for the memory feature. A value supplied by the app always wins.
pub(crate) fn memory_for_mode(
    mode: ClientMode,
    supplied: Option<MemoryConfiguration>,
) -> Option<MemoryConfiguration> {
    match supplied {
        Some(config) => Some(config),
        None if mode == ClientMode::Empty => Some(MemoryConfiguration::disabled()),
        None => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn tool_set_emits_source_qualified_patterns() {
        let v = ToolSet::new()
            .add_builtin("bash")
            .unwrap()
            .add_builtin("*")
            .unwrap()
            .add_custom("foo")
            .unwrap()
            .add_custom("*")
            .unwrap()
            .add_mcp("github-list_issues")
            .unwrap()
            .add_mcp("*")
            .unwrap()
            .to_vec();
        assert_eq!(
            v,
            vec![
                "builtin:bash",
                "builtin:*",
                "custom:foo",
                "custom:*",
                "mcp:github-list_issues",
                "mcp:*",
            ]
        );
    }

    #[test]
    fn tool_set_add_builtin_many() {
        let v = ToolSet::new()
            .add_builtin_many(BUILTIN_TOOLS_ISOLATED)
            .unwrap()
            .into_vec();
        assert_eq!(v.len(), BUILTIN_TOOLS_ISOLATED.len());
        assert_eq!(v[0], format!("builtin:{}", BUILTIN_TOOLS_ISOLATED[0]));
    }

    #[test]
    fn tool_set_rejects_invalid_names() {
        for bad in ["bash!", "with space", "colon:name", "", "wild*card"] {
            assert!(
                ToolSet::new().add_builtin(bad).is_err(),
                "expected '{bad}' to be rejected"
            );
            assert!(ToolSet::new().add_custom(bad).is_err());
            assert!(ToolSet::new().add_mcp(bad).is_err());
        }
    }

    #[test]
    fn tool_set_accepts_wildcard_and_underscores_and_dashes() {
        assert!(ToolSet::new().add_builtin("*").is_ok());
        assert!(ToolSet::new().add_mcp("github-list_issues").is_ok());
        assert!(ToolSet::new().add_custom("A_b-9").is_ok());
    }

    #[test]
    fn into_vec_is_idempotent_with_to_vec() {
        let ts = ToolSet::new().add_builtin("bash").unwrap();
        assert_eq!(ts.to_vec(), vec!["builtin:bash"]);
        assert_eq!(ts.into_vec(), vec!["builtin:bash"]);
    }

    #[test]
    fn into_vec_string_conversion() {
        let v: Vec<String> = ToolSet::new().add_mcp("*").unwrap().into();
        assert_eq!(v, vec!["mcp:*"]);
    }

    #[test]
    fn validate_tool_filter_list_rejects_bare_star() {
        let bad = vec!["*".to_string()];
        assert!(validate_tool_filter_list("availableTools", Some(&bad)).is_err());
    }

    #[test]
    fn validate_tool_filter_list_allows_qualified_star() {
        let ok = vec!["builtin:*".to_string(), "mcp:*".to_string()];
        assert!(validate_tool_filter_list("availableTools", Some(&ok)).is_ok());
    }

    #[test]
    fn validate_tool_filter_list_none_is_ok() {
        assert!(validate_tool_filter_list("availableTools", None).is_ok());
    }

    #[test]
    fn builtin_tools_isolated_contents() {
        assert!(BUILTIN_TOOLS_ISOLATED.contains(&"ask_user"));
        assert!(BUILTIN_TOOLS_ISOLATED.contains(&"task_complete"));
        assert!(BUILTIN_TOOLS_ISOLATED.contains(&"skill"));
        assert!(!BUILTIN_TOOLS_ISOLATED.contains(&"bash"));
        assert!(!BUILTIN_TOOLS_ISOLATED.contains(&"edit"));
        assert!(!BUILTIN_TOOLS_ISOLATED.contains(&"web_fetch"));
    }

    #[test]
    fn client_mode_default_is_copilot_cli() {
        assert_eq!(ClientMode::default(), ClientMode::CopilotCli);
    }

    #[test]
    fn system_message_copilot_cli_passes_through_unchanged() {
        let cfg = SystemMessageConfig {
            mode: Some("append".to_string()),
            content: Some("hello".to_string()),
            sections: None,
        };
        let out = system_message_for_mode(ClientMode::CopilotCli, Some(cfg.clone()));
        let out = out.unwrap();
        assert_eq!(out.mode.as_deref(), Some("append"));
        assert_eq!(out.content.as_deref(), Some("hello"));
    }

    #[test]
    fn system_message_empty_none_injects_strip() {
        let out = system_message_for_mode(ClientMode::Empty, None).unwrap();
        assert_eq!(out.mode.as_deref(), Some("customize"));
        let sections = out.sections.unwrap();
        let env = sections.get("environment_context").unwrap();
        assert_eq!(env.action.as_deref(), Some("remove"));
    }

    #[test]
    fn system_message_empty_append_promoted_to_customize() {
        let cfg = SystemMessageConfig {
            mode: Some("append".to_string()),
            content: Some("hi".to_string()),
            sections: None,
        };
        let out = system_message_for_mode(ClientMode::Empty, Some(cfg)).unwrap();
        assert_eq!(out.mode.as_deref(), Some("customize"));
        assert_eq!(out.content.as_deref(), Some("hi"));
        let sections = out.sections.unwrap();
        assert!(sections.contains_key("environment_context"));
    }

    #[test]
    fn system_message_empty_replace_passes_through() {
        let cfg = SystemMessageConfig {
            mode: Some("replace".to_string()),
            content: Some("verbatim".to_string()),
            sections: None,
        };
        let out = system_message_for_mode(ClientMode::Empty, Some(cfg.clone())).unwrap();
        assert_eq!(out.mode.as_deref(), Some("replace"));
        assert_eq!(out.content.as_deref(), Some("verbatim"));
        assert!(out.sections.is_none());
    }

    #[test]
    fn system_message_empty_customize_with_env_context_preserved() {
        let mut sections = HashMap::new();
        sections.insert(
            "environment_context".to_string(),
            SectionOverride {
                action: Some("replace".to_string()),
                content: Some("custom env".to_string()),
            },
        );
        let cfg = SystemMessageConfig {
            mode: Some("customize".to_string()),
            content: None,
            sections: Some(sections),
        };
        let out = system_message_for_mode(ClientMode::Empty, Some(cfg)).unwrap();
        let env = out.sections.unwrap().remove("environment_context").unwrap();
        assert_eq!(env.action.as_deref(), Some("replace"));
        assert_eq!(env.content.as_deref(), Some("custom env"));
    }

    #[test]
    fn system_message_empty_customize_without_env_context_gets_strip() {
        let mut sections = HashMap::new();
        sections.insert(
            "other_section".to_string(),
            SectionOverride {
                action: Some("replace".to_string()),
                content: Some("body".to_string()),
            },
        );
        let cfg = SystemMessageConfig {
            mode: Some("customize".to_string()),
            content: None,
            sections: Some(sections),
        };
        let out = system_message_for_mode(ClientMode::Empty, Some(cfg)).unwrap();
        let secs = out.sections.unwrap();
        assert!(secs.contains_key("other_section"));
        let env = secs.get("environment_context").unwrap();
        assert_eq!(env.action.as_deref(), Some("remove"));
    }

    #[test]
    fn memory_copilot_cli_leaves_unset_when_not_supplied() {
        assert_eq!(memory_for_mode(ClientMode::CopilotCli, None), None);
    }

    #[test]
    fn memory_copilot_cli_preserves_supplied() {
        assert_eq!(
            memory_for_mode(ClientMode::CopilotCli, Some(MemoryConfiguration::enabled())),
            Some(MemoryConfiguration::enabled())
        );
    }

    #[test]
    fn memory_empty_defaults_to_disabled() {
        assert_eq!(
            memory_for_mode(ClientMode::Empty, None),
            Some(MemoryConfiguration::disabled())
        );
    }

    #[test]
    fn memory_empty_preserves_supplied() {
        assert_eq!(
            memory_for_mode(ClientMode::Empty, Some(MemoryConfiguration::enabled())),
            Some(MemoryConfiguration::enabled())
        );
    }
}
